### ISSUE TITLE
Help understanding logits and model vocabs

### ISSUE THREAD

**Issue by thekevinscott** [xenova/transformers.js#713]
### Question

I'm trying to write a custom `LogitsProcessor` and have some questions. For reference, I'm using [`Xenova/phi-1_5_dev`](https://huggingface.co/Xenova/phi-1_5_dev). I'm trying to implement a custom logic for white or blacklisting tokens, but running into difficulties understanding how to interpret token ids, tokens, and their decoded counterparts.

Here's what I think I understand:

- [The vocab file is defined at `vocab.json`](https://huggingface.co/Xenova/phi-1_5_dev/blob/main/vocab.json), and has 50,257 entries. 
- This file is exposed on `pipeline.tokenizer.vocab`, translated from the object representation of `vocab.json` (`{ token: tokenID }`), to an array of `token`s whose indices correspond to `tokenID`. 
  - **Question:** `vocab.json` has 50,257 entries, but `pipeline.tokenizer.vocab` has 50,295 entries. Is this because `pipeline.tokenizer.vocab` _also_ includes `added_tokens.json`?
  - And [`special_tokens_map.json`](https://huggingface.co/Xenova/phi-1_5_dev/blob/main/special_tokens_map.json) is already included in `vocab.json` it appears
- The tokens in the vocab file must be decoded before being displayed
  - for example, the token in `vocab.json` at `50255` is `"Ġgazed"`, but if I decode this character by character (`pipeline.tokenizer.decoder.byte_decoder('Ġ')` becomes `32` which corresponds to a space `" "`) I get `" gazed"`. I _think_ these correspond to code points.
- The `logits` argument contains scores where the index of each score is the `tokenID`. So setting the score at position `50255` to `-Infinity` should ensure that the token `"Ġgazed"` (or, decoded, `" gazed"`) should never appear.
- The `logits` argument I'm getting back for this model in my `LogitsProcessor` has dimensions of `[51200,]`. `pipeline.tokenizer.vocab` has size of is 50,295. That would seem to indicate 905 unused tokens at the end of the tensor; can these be safely ignored, or do they correspond to something important that I'm missing?

I'd appreciate any insight or feedback on whether my assumptions above are correct or not. Thank you!

**Comment by thekevinscott** [comment]
The precipitating reason for this question is that I'm trying to force an `eos` token, but the model continues generating.

Here's a sample `LogitsProcessor`:

```
class LogitsProcessor {
  pipeline: TextGenerationPipeline;
  tokenIds: number[];
  stopTokenId: number;

  constructor(pipeline: TextGenerationPipeline, str: string) {
    this.pipeline = pipeline;
    const { input_ids, } = (pipeline.tokenizer as TokenizeFn)(str);
    this.tokenIds = [...input_ids.data,].map((n: bigint) => Number(n));
    this.stopTokenId = this.pipeline.tokenizer.model.convert_tokens_to_ids([
      this.pipeline.tokenizer.getToken('eos_token'),
    ])[0];
  }

  processors = [(inputTokens: number[], logits: Tensor) => {
    if (inputTokens.length > this.tokenIds.length) {
      console.warn('This should not happen');
      return logits;
    }
    const tokenId = this.tokenIds[inputTokens.length];
    const id = inputTokens.length === this.tokenIds.length ? this.stopTokenId : this.tokenIds[inputTokens.length];
    logits.data.fill(-Infinity);
    logits.data[id] = Infinity;
    return logits;
  },];

  [Symbol.iterator]() {
    return this.processors.values();
  }
}
```

instantiated with:

```
    const prompt = 'Write me some code';
    const logitsProcessor = new LogitsProcessor(this.pipeline, prompt + ' foo');
```

This will generate output like the following:

```
Write me some code foo<|endoftext|>
Student: A company has a budget of $5000 to spend on advertising. They want
```

I don't understand why `<|endoftext|>` is being treated like a part of the text output and not an indication to stop generation. I assume the answer is because I'm not understanding something in my initial question above.

**Comment by thekevinscott** [comment]
Hmm. I just tried with `Xenova/gpt2` and I now see the following output:

```
Write me some code foo<|endoftext|>
```

So, _maybe_ the issue with the model not stopping is specific to the model being used?

I still don't understand why `<|endoftext|>` is being returned as part of the text generated, though.

**Comment by thekevinscott** [comment]
> I still don't understand why <|endoftext|> is being returned as part of the text generated, though.

This was user error. I'm calling `.decode()` manually and was neglecting to pass `skip_special_tokens`; this omits the `<|endoftext|>` from the text:

```
    const decoded = tokenizer.decode(outputTokenIds[0], {
      skip_special_tokens: true,
    });
```

I still don't understand why `phi-1_5_dev` is not stopping on an eos token though.

**Comment by thekevinscott** [comment]
Here's what I've found for `phi-1_5_dev`:

- `pipeline.model.config.eos_token_id` is `2`
- `pipeline.tokenizer.model.convert_tokens_to_ids([pipeline.tokenizer.getToken('eos_token'),])[0]` is `50256`

Returning `2` as the eos token successfully stops generation for `phi-1_5_dev`. However, the token gets incorrectly decoded:

```
    const decoded = tokenizer.decode(outputTokenIds[0], {
      skip_special_tokens: true,
    });

> Write me some code foo#
```

Which makes sense, as `2` is `#` in the `vocab.json`.

So I guess this entire thread boils down to: why the discrepancy?

**Comment by xenova** [comment]
Thanks for the report! Indeed, it looks like the EOS token is incorrectly set in the original version of the model (see [here](https://huggingface.co/susnato/phi-1_5_dev/blob/main/config.json#L6)), which is why it's also (incorrectly) set to 2 in our version. I have updated it [here](https://huggingface.co/Xenova/phi-1_5_dev/commit/e90890a0c95076e3fc6185d11454c14ee902f211) (it's set to `null` [here](https://huggingface.co/microsoft/phi-1_5/blob/main/config.json#L13)).

**Comment by xenova** [comment]
As for your original question, you can use the `NoBadWordsLogitsProcessor` logits processor for this (see [here](https://github.com/xenova/transformers.js/blob/642743136efa3a481b2b96d3c2c550085540d844/src/utils/generation.js#L496))

**Comment by thekevinscott** [comment]
Thanks for the response! Those edits look great. Does `eos_token_id: null` imply anything in particular, or does it just mean it falls because to whatever the default `eos_token_id` is?

> As for your original question, you can use the NoBadWordsLogitsProcessor logits processor for this (see [here](https://github.com/xenova/transformers.js/blob/642743136efa3a481b2b96d3c2c550085540d844/src/utils/generation.js#L496))

Appreciate that reference. My use case is a bit more complicated - I'm trying to implement a GBNF grammar parser similar to `llama.cpp`'s implementation. But good to know this exists so I don't have to reinvent the wheel in the future!
